### DISCUSSION TITLE
Default Optimization (TPE) does NOT improve the last-best-trial (reproducible code attached). Parameter setting suggestions or bugs?

### DISCUSSION THREAD

**Discussion by pangjac** [optuna/optuna#5114] [Q&A]
Hi,

I am trying a toy regression problem, where the evaluation metric is to let the total percentage of y_pred under a certain range `(-inf, Lower_bound]` as higher possible (see below code `eval_metric_accuracy_customized`). Noticed that for this regression dataset, the total percentage of y_value <0 is of 50% (0.5).

```
import numpy as np
import optuna

import sklearn.datasets
import sklearn.metrics
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_regression
import xgboost as xgb

# make a random regression problem dataset
X, y = make_regression(n_samples=1000, n_features=15, noise=1, random_state=42)
train_x, valid_x, train_y, valid_y = train_test_split(X, y, test_size=0.25, random_state=42)
dtrain = xgb.DMatrix(train_x, label=train_y)
dvalid = xgb.DMatrix(valid_x, label=valid_y)
print(pd.DataFrame(y).describe())

# define our customized metric function
def eval_metric_accuracy_customized(y_pred, y_true, sample_weight=None, LB=0, UB=100) :
    #if y_true <=LB and corresponding y_pred <= (1+threshold)*LB, 
    # we consider this prediction is accurate.
    cond1 = (y_pred <= (1+threshold)*LB) & (y_true <= LB)
    calc_weight= np.where(cond1, 1.0, 0.0)
    acc = np.mean(calc_weight)
    # acc indicate the proportion of y_pred fall within the condition of cond1
    return acc


# in idea world where the y_pred is exactly the same as y_true, below evaluation returns 0.5
eval_metric_accuracy_customized(y_pred=valid_y, y_true=valid_y, LB=0, UB=200)

np.mean(valid_y < 0) # 0.5 because actual y_true <0 is 50% of the whole valid_y 
```

I noticed that **using default TPE Sampler does not improve the last trial score**!  I was expecting that since TPE is Bayesian optimization using kernel fitting, at least I should expect some incremental improvements from trial to trial. However, **the best score trial is trial 1, and for the following 998 trials, no improvement at all** ( it is even worse).

Here is the code I use. I made a random data sample for a regression problem.

```
# now starts tuning 
def objective_internal(trial, dtrain=dtrain, dvalid=dvalid):
    
    param = {
        "objective": "reg:absoluteerror",  # we use mae as xgboost objective function.
        # use exact for small dataset.
        "tree_method": "exact",
        # defines booster, gblinear for linear functions.
        "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
        # L2 regularization weight.
        "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log=True),
        # L1 regularization weight.
        "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
        # sampling ratio for training data.
        "subsample": trial.suggest_float("subsample", 0.2, 1.0),
        # sampling according to each tree.
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.2, 1.0),
    }

    if param["booster"] in ["gbtree", "dart"]:
        # maximum depth of the tree, signifies complexity of the tree.
        param["max_depth"] = trial.suggest_int("max_depth", 3, 9, step=2)
        # minimum child weight, larger the term more conservative the tree.
        param["min_child_weight"] = trial.suggest_int("min_child_weight", 2, 10)
        param["eta"] = trial.suggest_float("eta", 1e-8, 1.0, log=True)
        # defines how selective algorithm is.
        param["gamma"] = trial.suggest_float("gamma", 1e-8, 1.0, log=True)
        param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])

    if param["booster"] == "dart":
        param["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
        param["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
        param["rate_drop"] = trial.suggest_float("rate_drop", 1e-8, 1.0, log=True)
        param["skip_drop"] = trial.suggest_float("skip_drop", 1e-8, 1.0, log=True)

    bst = xgb.train(param, dtrain, verbose_eval=True)
    preds = bst.predict(dvalid)
    acc = eval_metric_accuracy_customized(y_true=valid_y, y_pred=preds,  LB=0, UB=120)
    print(f"custimozed acc = {acc}")
    return acc


# to avoid initialize dataset every time study call optimize function, saperate dataset out side of objective function.
from functools import partial
def objective_wrapper(dtrain, dvalid):
    return partial(objective_internal, dtrain = dtrain, dvalid=dvalid)

objective = objective_wrapper(dtrain, dvalid)

# tune!
from optuna.samplers import TPESampler
study = optuna.create_study(direction="maximize", sampler=TPESampler(n_startup_trials=100)) # maximize the accuracy is our tuning direction.
study.optimize(objective, n_trials=1000, timeout=600)

```

A sample tuning log I have is below, where you will find that `Best is trial 1 with value: 0.12.` . 

**Two issues:
1. best trial value `0.12` is far from the `0.5` ( in ideal world where the y_pred is exactly the same as y_true, the evaluation value is 0.5)
2. the best trial value `0.12` achieves at the trial 1, and it never improves for the following 998 trials. What's worse is that for all the following trials, some trial even returns accuracy 0.0
**
```
[I 2023-11-14 18:03:45,999] A new study created in memory with name: no-name-2fccd619-5fb7-4b8b-8047-5067508a61f2
[I 2023-11-14 18:03:46,102] Trial 0 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 5.129650812419563e-05, 'alpha': 1.520969275647894e-08, 'subsample': 0.4598227640770435, 'colsample_bytree': 0.8313522382964404, 'max_depth': 3, 'min_child_weight': 2, 'eta': 0.0001869710758563331, 'gamma': 0.6169164663508273, 'grow_policy': 'depthwise'}. Best is trial 0 with value: 0.0.
[I 2023-11-14 18:03:46,112] Trial 1 finished with value: 0.12 and parameters: {'booster': 'gblinear', 'lambda': 0.00010601070472641276, 'alpha': 3.7550963969497754e-06, 'subsample': 0.7803228327801253, 'colsample_bytree': 0.7601824404069766}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,123] Trial 2 finished with value: 0.116 and parameters: {'booster': 'gblinear', 'lambda': 4.6359515749754716e-07, 'alpha': 0.0030297985248756015, 'subsample': 0.8073590526799577, 'colsample_bytree': 0.48939212841808283}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,234] Trial 3 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 8.923696072410334e-05, 'alpha': 0.001593375924920331, 'subsample': 0.5200532178301664, 'colsample_bytree': 0.7101168428893736, 'max_depth': 7, 'min_child_weight': 5, 'eta': 0.00025306750317905267, 'gamma': 3.026696920885293e-08, 'grow_policy': 'lossguide'}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,242] Trial 4 finished with value: 0.0 and parameters: {'booster': 'gblinear', 'lambda': 0.5104677459456162, 'alpha': 0.01965637209927828, 'subsample': 0.7731146965598847, 'colsample_bytree': 0.7725664478736085}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,250] Trial 5 finished with value: 0.06 and parameters: {'booster': 'gblinear', 'lambda': 1.0209211825610941e-07, 'alpha': 0.06686152400085864, 'subsample': 0.48996373267791715, 'colsample_bytree': 0.20099126338570195}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,259] Trial 6 finished with value: 0.116 and parameters: {'booster': 'gblinear', 'lambda': 4.397061315795294e-05, 'alpha': 0.012557625988932087, 'subsample': 0.9116038236945725, 'colsample_bytree': 0.8028037389126541}. Best is trial 1 with value: 0.12.
custimozed acc = 0.0
custimozed acc = 0.12
custimozed acc = 0.116
custimozed acc = 0.0
custimozed acc = 0.0
custimozed acc = 0.06
custimozed acc = 0.116
[I 2023-11-14 18:03:46,423] Trial 7 finished with value: 0.0 and parameters: {'booster': 'dart', 'lambda': 0.0008011524632934181, 'alpha': 2.4612418129430516e-06, 'subsample': 0.28951075590639086, 'colsample_bytree': 0.7453167740428308, 'max_depth': 7, 'min_child_weight': 6, 'eta': 0.17948308068471225, 'gamma': 3.302335223553575e-07, 'grow_policy': 'lossguide', 'sample_type': 'weighted', 'normalize_type': 'tree', 'rate_drop': 4.850368439274315e-06, 'skip_drop': 0.0003973068280998373}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,542] Trial 8 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 1.5078956506999752e-05, 'alpha': 0.003972741482350864, 'subsample': 0.8111496280497457, 'colsample_bytree': 0.5751340376939773, 'max_depth': 5, 'min_child_weight': 3, 'eta': 0.09279398423936037, 'gamma': 0.021454541962588038, 'grow_policy': 'depthwise'}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,555] Trial 9 finished with value: 0.088 and parameters: {'booster': 'gblinear', 'lambda': 0.012556559708914472, 'alpha': 0.024446675216426777, 'subsample': 0.5103771722790406, 'colsample_bytree': 0.3498495955093456}. Best is trial 1 with value: 0.12.
custimozed acc = 0.0
custimozed acc = 0.0
custimozed acc = 0.088
[I 2023-11-14 18:03:46,660] Trial 10 finished with value: 0.0 and parameters: {'booster': 'dart', 'lambda': 1.2201060066881958e-05, 'alpha': 0.005517343579761779, 'subsample': 0.21421151677943479, 'colsample_bytree': 0.9846739931251418, 'max_depth': 9, 'min_child_weight': 4, 'eta': 2.6550186546680567e-05, 'gamma': 0.014494623284744552, 'grow_policy': 'depthwise', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 0.0007378976631929169, 'skip_drop': 0.08584610458922132}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,780] Trial 11 finished with value: 0.0 and parameters: {'booster': 'dart', 'lambda': 2.801584646862098e-08, 'alpha': 0.3664121116790731, 'subsample': 0.36843126709408325, 'colsample_bytree': 0.7290181602778958, 'max_depth': 5, 'min_child_weight': 3, 'eta': 3.311683153164776e-08, 'gamma': 5.957000235702763e-08, 'grow_policy': 'depthwise', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 1.7672023570373324e-05, 'skip_drop': 0.0026673069508628532}. Best is trial 1 with value: 0.12.
custimozed acc = 0.0
custimozed acc = 0.0
[I 2023-11-14 18:03:46,895] Trial 12 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 4.130387427745429e-08, 'alpha': 2.5573774709936577e-07, 'subsample': 0.5457812690042174, 'colsample_bytree': 0.4472725571173326, 'max_depth': 9, 'min_child_weight': 10, 'eta': 0.0006267173668632139, 'gamma': 5.849619385001541e-08, 'grow_policy': 'depthwise'}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,906] Trial 13 finished with value: 0.116 and parameters: {'booster': 'gblinear', 'lambda': 9.738141060203865e-08, 'alpha': 0.0032043156499698324, 'subsample': 0.3051123474279007, 'colsample_bytree': 0.7167606445238244}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:46,987] Trial 14 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 7.443868387921888e-07, 'alpha': 3.857749360698529e-07, 'subsample': 0.9403936633937182, 'colsample_bytree': 0.9705019066035037, 'max_depth': 5, 'min_child_weight': 10, 'eta': 1.4999519878455846e-07, 'gamma': 0.10194584579540727, 'grow_policy': 'lossguide'}. Best is trial 1 with value: 0.12.
custimozed acc = 0.0
custimozed acc = 0.116
custimozed acc = 0.0
[I 2023-11-14 18:03:47,107] Trial 15 finished with value: 0.0 and parameters: {'booster': 'dart', 'lambda': 0.0002383112080577705, 'alpha': 0.016462898347311575, 'subsample': 0.8264722117449204, 'colsample_bytree': 0.6365629096417953, 'max_depth': 5, 'min_child_weight': 7, 'eta': 1.1264746758898723e-08, 'gamma': 0.00012088137286037438, 'grow_policy': 'lossguide', 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.8592544700539254e-08, 'skip_drop': 5.713950059206531e-05}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:47,181] Trial 16 finished with value: 0.0 and parameters: {'booster': 'gbtree', 'lambda': 0.028204633987629346, 'alpha': 0.9724339744645314, 'subsample': 0.8949528027281288, 'colsample_bytree': 0.7875474884967182, 'max_depth': 5, 'min_child_weight': 8, 'eta': 1.8157887144928414e-06, 'gamma': 1.0015350601683937e-07, 'grow_policy': 'depthwise'}. Best is trial 1 with value: 0.12.
[I 2023-11-14 18:03:47,267] Trial 17 finished with value: 0.0 and parameters: {'booster': 'dart', 'lambda': 0.0001376190175023213, 'alpha': 6.628528115390551e-07, 'subsample': 0.40934825243918893, 'colsample_bytree': 0.907733043824821, 'max_depth': 5, 'min_child_weight': 10, 'eta': 1.0442288221701071e-05, 'gamma': 0.0702242686420645, 'grow_policy': 'depthwise', 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 3.4689920659943325e-06, 'skip_drop': 0.08248129323831446}. Best is trial 1 with value: 0.12.
custimozed acc = 0.0
custimozed acc = 0.0
custimozed acc = 0.0

```
